src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/FormatterServices.cs

```diff
@@ -0,0 +1,299 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Globalization;
+using System.Diagnostics;
+using System.Runtime.Serialization.Formatters;
+
+namespace System.Runtime.Serialization
+{
+    public static class FormatterServices
+    {
+        private static readonly ConcurrentDictionary<MemberHolder, MemberInfo[]> s_memberInfoTable = new ConcurrentDictionary<MemberHolder, MemberInfo[]>();
+
+        private static FieldInfo[] GetSerializableFields(Type type)
+        {
+            if (type.GetTypeInfo().IsInterface)
+            {
+                return Array.Empty<FieldInfo>();
+            }
+
+            if (!type.GetTypeInfo().IsSerializable)
+            {
+                throw new SerializationException(SR.Format(SR.Serialization_NonSerType, type.GetTypeInfo().FullName, type.GetTypeInfo().Assembly.FullName));
+            }
+
+            var results = new List<FieldInfo>();
+            for (Type t = type; t != typeof(object); t = t.GetTypeInfo().BaseType)
+            {
+                foreach (FieldInfo field in t.GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly))
+                {
+                    if ((field.Attributes & FieldAttributes.NotSerialized) != FieldAttributes.NotSerialized)
+                    {
+                        results.Add(field);
+                    }
+                }
+            }
+            return results.ToArray();
+        }
+
+        public static MemberInfo[] GetSerializableMembers(Type type)
+        {
+            return GetSerializableMembers(type, new StreamingContext(StreamingContextStates.All));
+        }
+
+        public static MemberInfo[] GetSerializableMembers(Type type, StreamingContext context)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            // If we've already gathered the members for this type, just return them.
+            // Otherwise, get them and add them.
+            var mh = new MemberHolder(type, context);
+            MemberInfo[] members;
+            if (!s_memberInfoTable.TryGetValue(mh, out members))
+            {
+                members = GetSerializableFields(type);
+                s_memberInfoTable.TryAdd(mh, members);
```

Comment:
Why not use GetOrAdd?  Since the parallel evaluators should come up with the same answer it shouldn't really matter, but I was confused as to why you weren't checking the result of TryAdd and then got differently confused.